### PR TITLE
Implement Transformation JMES Path Predicates

### PR DESCRIPTION
**Pull request by AndreKurait**
### Description
Implement Transformation JMES Path Predicates

Components:
* Adds IJsonPredicate/Provider to define a spec for a custom precondition
* Adds JMESPathPredicate/Provider to define defining preconditions with JMES
* Adds PredicatesLoader to use service loading for preconditions
* Adds JsonConditionalTransformerProvider which uses PredicateLoader and TransformationLoader to pull in classes to delegate functionality to
* Separates Loaders into dedicated module `jsonMessageTransformerLoaders`
___
 
* Category: New Feature
* Why these changes are required? Improve Transformation Ux
* What is the old behavior before changes and new behavior after changes? New ability to specify conditional transformations.

### Issues Resolved
[MIGRATIONS-2146](https://opensearch.atlassian.net/browse/MIGRATIONS-2146)

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Added unit testing, performed some E2E manual testing in docker.

Example Transformation:

```json
[
  {
    "JsonConditionalTransformerProvider": [
      [
        {
          "JsonJMESPathPredicateProvider": {
            "script": "URI == '/testindex/_search'"
          }
        }
      ],
      [
        {
          "JsonJoltTransformerProvider": {
            "canned": "ADD_GZIP"
          }
        }
      ]
    ]
  },
  {
    "JsonJoltTransformerProvider": {
      "script": {
        "operation": "modify-overwrite-beta",
        "spec": {
          "headers": {
            "newHeader": "newValue"
          }
        }
      }
    }
  }
]
```

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by gregschohn — COMMENTED** [review]
Please add the input to what goes into the transformer.  I'd like future readers to see where URI is defined and to understand the general structure...

I have some significant concerns as to the approach now that I'm carefully going through this and thinking of the dev/user experience.  Some of the concerns are issues for the next corner (but maybe the very next corner, w/ genAI and stable ids coming into the mix), but it's useful to consider them to think about how this routing feature will interplay as the solution matures.

Concerns
1. We run transformations w/in a netty event loop.  Today, transformations are assumed to be non-blocking and non-preemptive.  How would we make them, at least in some modes, run asynchronously.
2. Today, we pass a java map into a transformation - tomorrow it may need to be something more like a stream (File Channel, etc).  I suspect that we’ll have a few layers of an onion here to help to facilitate between java-json interfaces and say purely text ones (Python, etc)
3. Predicates shouldn’t be able to make changes to the input.  It will be a miserable user experience if a predicate DOES make a change (inadvertently) and the next body or predicate is thrown off.

W/ those future changes, how can we minimize their impact to basic high-level control flow?  We need to eventually make our interfaces readonly/text/streams…
If we did that, setting up lots of separate parses (potentially) within each predicate function would be wasteful.

I can easily get the parse-passes down to two if I have a Document -> Label function (like a switch statement), but I’m not sure that I can do better w/out making the interfaces much more difficult.  This is a tradeoff between the complexity of pulling everything together and defining things in a limited number of places.  Running a whole bunch of boolean predicates w/ an entire document could be really expensive for the utility that users are getting.  It might also be a lot more for them to manage too - could it be easier if they just defined one routing transform?

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/JsonCompositePrecondition.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public boolean evaluatePrecondition(Map<String, Object> incomingJson) {
+        var preconditions = jsonPreconditionList.stream();
+        Predicate<IJsonPrecondition> predicate = p -> p.evaluatePrecondition(incomingJson);
+        switch (operation) {
+            case ALL:
+                return preconditions.allMatch(predicate);
+            case ANY:
+                return preconditions.anyMatch(predicate);
+            case NONE:
+                return preconditions.noneMatch(predicate);
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/JsonCompositePrecondition.java:35]: We don't have it now, but if we had JOLT Preconditions, then you could chain on both of them needing to hold.

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonJMESPathMessageTransformerProvider/src/main/java/org/opensearch/migrations/transform/JsonJMESPathPreconditionProvider.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            transformers.toArray(IJsonPrecondition[]::new));
+    }
+
+    private String getConfigUsageStr() {
+        return this.getClass().getName()
+            + " expects the incoming configuration "
+            + "to be a Map<String,Object> or a List<Map<String,Object>>.  "
+            + "Each of the Maps should have one key-value of \"script\": \"...\".  "
+            + "Script values should be a fully-formed inlined JsonPath queries encoded as a json value.  "
+            + "All of the values within a configuration will be concatenated into one chained precondition.";
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonJMESPathMessageTransformerProvider/src/main/java/org/opensearch/migrations/transform/JsonJMESPathPreconditionProvider.java:58]: We create a JsonCompositePrecondition with a CompositeOperation of ALL. so if a list is passed in, we'll check the AND of all the preconditions

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonJMESPathMessageTransformer/src/main/java/org/opensearch/migrations/transform/JsonJMESPathPrecondition.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import io.burt.jmespath.BaseRuntime;
+import io.burt.jmespath.Expression;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class JsonJMESPathPrecondition implements IJsonPrecondition {
+
+    Expression<Object> expression;
+
+    public JsonJMESPathPrecondition(BaseRuntime<Object> runtime, String script) {
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonJMESPathMessageTransformer/src/main/java/org/opensearch/migrations/transform/JsonJMESPathPrecondition.java:14]: Are you saying to wrap the script as the value in a transformer and get the value back out?

File: TrafficCapture/trafficReplayer/build.gradle [hunk 4]
```diff
@@ -60,6 +60,7 @@ dependencies {
     testImplementation project(':TrafficCapture:transformationPlugins:jsonMessageTransformers:jsonJMESPathMessageTransformerProvider')
     testImplementation project(':TrafficCapture:transformationPlugins:jsonMessageTransformers:jsonJoltMessageTransformerProvider')
     testImplementation project(':TrafficCapture:transformationPlugins:jsonMessageTransformers:openSearch23PlusTargetTransformerProvider')
+    testImplementation project(':TrafficCapture:transformationPlugins:jsonMessageTransformers:jsonJoltMessageTransformer')
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply TrafficCapture/trafficReplayer/build.gradle:63]: Wasn't needed. I moved PayloadRepackingTest and AddCompressionEncoding test back to jsonJoltMessageTransformerProvider and removed this

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonPrecondition.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package org.opensearch.migrations.transform;
+
+import java.util.Map;
+
+
+public interface IJsonPrecondition {
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonPrecondition.java:6]: Renamed to Predicate

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonPrecondition.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+    boolean evaluatePrecondition(Map<String, Object> incomingJson);
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonPrecondition.java:7]: Are you meaning to have this be `boolean evaluate(...`? 
I intentionally differentiated the name to allow for multiple inheritance of interfaces in the case IJsonTransformer defined `Map<String, Object> evaluate(Map<String, Object> incomingJson);

I'll rename this to predicate

[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonPrecondition.java]: I think that this should follow a Predicate pattern.  IJsonPredicate::test(thing) -> boolean

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/JsonCompositePrecondition.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/JsonCompositePrecondition.java]: If we could have a jolt precondition, I'd want to redo this entire PR.
A jolt transform to change the input whereas a JMESPath one cannot

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonPrecondition.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (AndreKurait) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonPrecondition.java]: Updated with
```
public interface IJsonPredicate extends Predicate<Map<String, Object>>
```

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonJMESPathMessageTransformer/src/main/java/org/opensearch/migrations/transform/JsonJMESPathPrecondition.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonJMESPathMessageTransformer/src/main/java/org/opensearch/migrations/transform/JsonJMESPathPrecondition.java]: That was one thought, but I didn't like that one in the end.  Another option could be to define both of these methods in the same interface and use the same loading functionality for both.  I haven't thought of that too much.  I just didn't like seeing the parallel universes developing.

File: TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/JsonCompositePrecondition.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (AndreKurait) [reply TrafficCapture/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/JsonCompositePrecondition.java]: Let's put a pin on non-JMESPath predicates and restrict ourselves to this for the meantime

**Review by gregschohn — APPROVED** [review]

**Pull request merged**